source/Plugins/Language/Swift/SwiftCodeCompletion.cpp

```diff
@@ -192,8 +192,166 @@ doCodeCompletion(SourceFile &SF, StringRef EnteredCode,
   return BufferID;
 }
 
-CompletionResponse SwiftCompleteCode(SourceFile &SF, StringRef EnteredCode) {
-  ASTContext &Ctx = SF.getASTContext();
+static void AddSourceFile(ASTContext &Ctx, ModuleDecl *Module,
+                          SourceFileKind Kind) {
+  Optional<unsigned> BufferID;
+  SourceFile *SF = new (Ctx) SourceFile(
+      *Module, Kind, BufferID, SourceFile::ImplicitModuleImportKind::Stdlib,
+      /*Keep tokens*/ false);
+  SF->ASTStage = SourceFile::TypeChecked;
+  Module->addFile(*SF);
+}
+
+static SourceFile *GetSingleSourceFile(ModuleDecl *Module,
+                                       SourceFileKind Kind) {
+  SourceFile *Result = nullptr;
+  for (auto *File : Module->getFiles()) {
+    auto *SF = dyn_cast<SourceFile>(File);
+    if (!SF)
+      continue;
+    if (SF->Kind != Kind)
+      continue;
+    assert(!Result && "multiple source files of the requested kind");
+    Result = SF;
+  }
+  return Result;
+}
+
+CompletionResponse
+SwiftCompleteCode(SwiftASTContext &SwiftCtx,
+                  SwiftPersistentExpressionState &PersistentExpressionState,
+                  StringRef EnteredCode) {
+  Status Error;
+  ASTContext &Ctx = *SwiftCtx.GetASTContext();
+
+  // == Prepare a module and source files ==
+
+  // Get or create the module that we do completions in.
+  static ConstString CompletionsModuleName("completions");
+  ModuleDecl *CompletionsModule =
+      SwiftCtx.GetModule(CompletionsModuleName, Error);
+  if (!CompletionsModule) {
+    CompletionsModule = SwiftCtx.CreateModule(CompletionsModuleName, Error);
+    if (!CompletionsModule)
+      return CompletionResponse::error("could not make completions module");
+
+    // This file accumulates all of the "hand imports" (imports that the user
+    // made in previous executions). We also put the current entered code in
+    // this file.
+    AddSourceFile(Ctx, CompletionsModule, SourceFileKind::REPL);
+
+    // We reset this file with the persistent decls every completion request.
+    AddSourceFile(Ctx, CompletionsModule, SourceFileKind::Library);
+  }
+
+  // This file accumulates all of the "hand imports" (imports that the user
+  // made in previous executions). We also put the current entered code in
+  // this file.
+  SourceFile *EnteredCodeFile =
+      GetSingleSourceFile(CompletionsModule, SourceFileKind::REPL);
+  assert(EnteredCodeFile);
+
+  // Accumulate new hand imports into the file.
```

Comment:
I wonder if you can simply reuse SwiftExpressionParser::AutoImport?

Reply:
Also discussed a bit in person.

It sounds like a good goal to dedupe this with SwiftExpressionParser::AutoImport eventually, but it'll be a lot of work because SwiftExpressionParser::AutoImport does a lot of things and deals with a lot of state that might not be good for the completer to touch.

So we'll defer this until later.